### PR TITLE
Implementing Task / Context manager

### PR DESCRIPTION
**Pull request by aardschok**
Early draft of functioning tool
* Adds current context view in Avalon menu
* Adds contextmanager to tools library
* Set Context launcher contextmanager

### CODE REVIEW

File: avalon/tools/projectmanager/widget.py [hunk 1]
```diff
@@ -165,8 +165,8 @@ class AssetModel(TreeModel):
     Deprecated = 2
     ObjectId = 3
 
-    DocumentRole = QtCore.Qt.UserRole + 1
-    ObjectIdRole = QtCore.Qt.UserRole + 2
+    DocumentRole = QtCore.Qt.UserRole + 2
```
Comment (BigRoy) [inline avalon/tools/projectmanager/widget.py]: To describe why this has changed; This fixes it overriding the NodeRole as it did previously.

File: avalon/tools/projectmanager/widget.py [hunk 2]
```diff
@@ -271,19 +274,18 @@ def data(self, index, role):
...
 
         if role == self.ObjectIdRole:
-            node = index.internalPointer()
             return node.get("_id", None)
 
         if role == self.DocumentRole:
-            node = index.internalPointer()
             return node.get("_document", None)
 
+        if role == self.NodeRole:
```
Comment (BigRoy) [inline avalon/tools/projectmanager/widget.py]: This is wrong; remove this. :)

File: avalon/tools/projectmanager/widget.py [hunk 3]
```diff
@@ -237,10 +237,13 @@ def flags(self, index):
 
     def data(self, index, role):
 
+        if not index.isValid():
```
Comment (BigRoy) [inline avalon/tools/projectmanager/widget.py]: Perfect 👍

File: avalon/tools/projectmanager/widget.py [hunk 4]
```diff
@@ -237,10 +237,13 @@ def flags(self, index):
...
+            return
+
+        node = index.internalPointer()
```
Comment (BigRoy) [inline avalon/tools/projectmanager/widget.py]: Not too big of an issue, just note that this isn't actually **always** used. A lot of the calls to this function will pass all the way to the `super` call at the bottom.

**Comment by BigRoy** [comment]
Aside from the notes above this looks good - I think it's good to have a think about how this actually will get used in production.

So I have an open Maya session, and want to switch to another shot or asset to work on something else. I use this to set my context - what's next? How do I ensure I work correctly within that specific context? When releasing this tool to the artists it's good to elaborate a bit on how you intended it to be used and let's see in production if that misses any use cases. I'm also wondering how often this will actually get used - maybe artists are more likely to secretly just "CTRL+O" and open a file from elsewhere?

**Comment by aardschok** [comment]
I think we should let people refrain from using `CTRL+O` to switch context or tell them they can use it but only AFTER they have set the context correctly.
Most cases will be within the same asset but a different task, model to lookdev / model to rig / lookdev to setdress.

There might be a call to make this the launcher as you said earlier. We will need to see this in action.

**Comment by BigRoy** [comment]
> I think we should let people refrain from using CTRL+O to switch context or tell them they can use it but only AFTER they have set the context correctly.

This will work perfectly though - pipeline-wise there's currently no reason to now allow it. ;) Also, clicking a recent file will then also work; which could be convenient?

File: avalon/maya/pipeline.py [hunk 5]
```diff
@@ -102,6 +102,24 @@ def deferred():
...
+                                        api.Session["AVALON_TASK"])
+        context_menu = cmds.menuItem("currentContext",
+                                     label=context_label,
+                                     parent=self._menu,
+                                     subMenu=True)
+
+        cmds.menuItem("setCurrentContext",
+                      label="Set Context",
+                      parent=context_menu,
+                      command="print 'context: {}'".format(context_label))
```
Comment (BigRoy) [inline avalon/maya/pipeline.py]: This should still be changed to actually launch the tool! :)

**Pull request merged**